firedrake/slate/slate.py

```diff
@@ -240,62 +286,84 @@ def _output_string(self, prec=None):
         return ["S", "V", "M"][self.rank] + "_%d" % self.id
 
     def __repr__(self):
-        """SLATE representation of the tensor object."""
+        """Slate representation of the tensor object."""
         return ["Scalar", "Vector", "Matrix"][self.rank] + "(%r)" % self.form
 
+    @cached_property
     def _key(self):
         """Returns a key for hash and equality."""
         return (type(self), self.form)
 
 
-class UnaryOp(TensorBase):
-    """An abstract SLATE class for representing unary operations on a
-    Tensor object.
+class TensorOp(TensorBase):
+    """An abstract Slate class representing general operations on
+    existing Slate tensors.
 
-    :arg A: a :class:`TensorBase` object. This can be a terminal tensor object
-            (:class:`Tensor`) or any derived expression resulting from any number
-            of linear algebra operations on `Tensor` objects. For example,
-            another instance of a `UnaryOp` object is an acceptable input, or
-            a `BinaryOp` object.
+    :arg operands: an iterable of operands that are :class:`TensorBase`
+                   objects.
     """
 
-    def __init__(self, A):
-        """Constructor for the UnaryOp class."""
-        super(UnaryOp, self).__init__()
-        self.tensor = A
+    def __init__(self, *operands):
+        """Constructor for the TensorOp class."""
+        super(TensorOp, self).__init__()
+        self.operands = tuple(operands)
 
     def coefficients(self):
-        """Returns a tuple of coefficients associated with the tensor."""
-        return self.tensor.coefficients()
+        """Returns the expected coefficients of the resulting tensor."""
+        coeffs = [op.coefficients() for op in self.operands]
+        return tuple(OrderedDict.fromkeys(chain(*coeffs)))
 
     def ufl_domains(self):
         """Returns the integration domains of the integrals associated with
         the tensor.
         """
-        return self.tensor.ufl_domains()
+        collected_domains = [op.ufl_domain() for op in self.operands]
+        return join_domains(collected_domains)
```

Comment:
Does this want to be `join_domains(itertools.chain(*[op.ufl_domains() for op in self.operands]))` that DTRT if the operands have multiple domains, no?